sdk/resourcemanager/Azure.ResourceManager.Core/src/ApiVersions.cs

```diff
@@ -0,0 +1,112 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using Azure.ResourceManager.Resources;
+using System.Reflection;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Threading.Tasks;
+
+namespace Azure.ResourceManager.Core
+{
+    /// <summary>
+    /// A class representing Azure resource manager client options.
+    /// </summary>
+    public class ApiVersions
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ApiVersions"/> class.
+        /// </summary>
+        public ApiVersions()
+        {
+            BuildApiTable();
+        }
+
+        private Dictionary<string, string> _resourceToApiVersions = new Dictionary<string, string>();
+
+        private void BuildApiTable()
+        {
+            var methods = GetExtensionMethods();
+            foreach (var method in methods)
+            {
+                if (method.Name.EndsWith("RestApiVersions"))
+                {
+                    var apiObject = method.Invoke(null, new object[] { this });
+                    var properties = apiObject.GetType().GetProperties();
+                    foreach (var prop in properties)
+                    {
+                        if (typeof(ApiVersionsBase).IsAssignableFrom(prop.PropertyType))
+                        {
+                            var propVal = (ApiVersionsBase)prop.GetValue(apiObject);
+                            Console.WriteLine(prop.GetType());
+                            var key = propVal.ResourceType;
+                            _resourceToApiVersions.Add(key.ToString(), propVal);
+                        }
+                    }
+                }
+            }
+        }
+
+        private static IEnumerable<MethodInfo> GetExtensionMethods()
+        {
+            var results =
+                        from assembly in AppDomain.CurrentDomain.GetAssemblies()
+                        from type in assembly.GetTypes()
+                        where type.IsSealed && !type.IsGenericType && !type.IsNested
+                        from method in type.GetMethods(BindingFlags.Static | BindingFlags.Public)
+                        where method.IsDefined(typeof(ExtensionAttribute), false)
+                        where method.GetParameters()[0].ParameterType == typeof(ApiVersions)
+                        select method;
+            return results;
+        }
+
+        internal string LoadApiVersion(ProvidersOperations providers, ResourceIdentifier id)
+        {
+            var results = providers.Get(id.Type.Namespace);
```

Comment:
should first check if it was loaded between the time the genOp was constructed and the API call was made, otherwise you will be overriding the value here

Reply:
This only get's called on construction when a new ClientOptions is created. Are you saying if some other ClientOptions was made and that had values set then we should update those here?

For example
var options = new ArmClientOptions(); //Dictionary get's built once with default values
options.ComputeApiVersions.VirtualMachineVersion = VirtualMachineVersion.2020_06_01;//update

var options2 = new ArmClientOptions(); //We should default the Dictionary in this object to what it is set in options?